Add strict mypy type checking and update code_template.py

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42322 Add strict mypy type checking and update code_template.py**

Our current type checking rules are rather lax, and for
example don't force users to make sure they annotate all functions
with types.  For code generation code, it would be better to force
100% typing.  This PR introduces a new mypy configuration
mypy-strict.ini which applies rules from --strict.  We extend
test_type_hints.py to test for this case.  It only covers
code_template.py, which I have made strict clean in this PR.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22846120](https://our.internmc.facebook.com/intern/diff/D22846120)

ezyang committed:
Add strict mypy type checking and update code_template.py

Our current type checking rules are rather lax, and for
example don't force users to make sure they annotate all functions
with types.  For code generation code, it would be better to force
100% typing.  This PR introduces a new mypy configuration
mypy-strict.ini which applies rules from --strict.  We extend
test_type_hints.py to test for this case.  It only covers
code_template.py, which I have made strict clean in this PR.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add strict mypy type checking and update code_template.py"


Our current type checking rules are rather lax, and for
example don't force users to make sure they annotate all functions
with types.  For code generation code, it would be better to force
100% typing.  This PR introduces a new mypy configuration
mypy-strict.ini which applies rules from --strict.  We extend
test_type_hints.py to test for this case.  It only covers
code_template.py, which I have made strict clean in this PR.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22846120](https://our.internmc.facebook.com/intern/diff/D22846120)

[ghstack-poisoned]

ezyang committed:
Update on "Add strict mypy type checking and update code_template.py"


Our current type checking rules are rather lax, and for
example don't force users to make sure they annotate all functions
with types.  For code generation code, it would be better to force
100% typing.  This PR introduces a new mypy configuration
mypy-strict.ini which applies rules from --strict.  We extend
test_type_hints.py to test for this case.  It only covers
code_template.py, which I have made strict clean in this PR.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22846120](https://our.internmc.facebook.com/intern/diff/D22846120)

[ghstack-poisoned]

rgommers commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1536,10 +1536,15 @@ def process_legacy_th_option(option):
+            # technically isn't a Mapping, as it doesn't implement
+            # enough methods.  I could fix this with a Protocol but
+            # then I need typing_extensions which isn't currently
+            # a build dep.
```
That's the second time today `typing_extensions` would've come in handy. I opened https://github.com/pytorch/pytorch/issues/38221 for adding it a while ago, let me know if you want me to open a PR for that.

rgommers commented on mypy-strict.ini:
```diff
@@ -0,0 +1,33 @@
+disallow_any_unimported = True
+
+# Across versions of mypy, the flags toggled by --strict vary.  To ensure
+# we have reproducible type check, we instead manualy specify the flags
```
nit: typo in `manualy`

rgommers commented on mypy-strict.ini:
```diff
@@ -0,0 +1,33 @@
+warn_redundant_casts = True
+warn_unused_ignores = True
+warn_return_any = True
+no_implicit_reexport = True
```
nit: the documented `.ini` option is `implicit_reexport`, so changing to `implicit_reexport = False` may be nice

rgommers commented on test/test_type_hints.py:
```diff
@@ -204,12 +204,38 @@ def is_torch_mypyini(path_to_file):
 
         cwd = os.getcwd()
+        # TODO: Would be better not to chdir here, this affects the entire
+        # process!
```
good point adding the `try`-`except`. not sure the `TODO` is needed here, this was hard to avoid and it seems harmless - parallel test execution with `pytest-xdist` should use separate processes, and that's the one relevant case I can think of.

ezyang replied:
It's just a general hygiene thing, from my days when I could actually run thread concurrently :) You could trigger this via just the threading module, and people do use it in Python. OTOH I agree that it is relatively harmless and I'm not going to go out of my way to try to fix it, especially if mypy doesn't support explicitly specifying what cwd to run in!

malfet replied:
According to mypy developers, it's not supposed to throw, but I guess it's a good hygiene

rgommers reviewed (approved):
LGTM, just a few small comments

ezyang committed:
Update on "Add strict mypy type checking and update code_template.py"


Our current type checking rules are rather lax, and for
example don't force users to make sure they annotate all functions
with types.  For code generation code, it would be better to force
100% typing.  This PR introduces a new mypy configuration
mypy-strict.ini which applies rules from --strict.  We extend
test_type_hints.py to test for this case.  It only covers
code_template.py, which I have made strict clean in this PR.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D22846120](https://our.internmc.facebook.com/intern/diff/D22846120)

[ghstack-poisoned]
